Traduction de la page "Rendering Elements" (#17)
reactjs/fr.reactjs.org: content/docs/rendering-elements.md

---
id: rendering-elements
title: Rendu d’éléments
permalink: docs/rendering-elements.html
redirect_from:
  - "docs/displaying-data.html"
prev: introducing-jsx.html
next: components-and-props.html
---

Les éléments constituent les plus petits blocs d’une application React.

Un élément retranscrit ce que vous voulez voir à l’écran :

```js
const element = <h1>Bonjour, monde</h1>;
```

Contrairement aux élements DOM d’un navigateur, les éléments React sont de simples objets peu coûteux à créer. Le DOM React se charge de mettre à jour le DOM afin qu’il corresponde aux éléments React.

>**Note :**
>
>On pourrait confondre les éléments avec un concept plus largement connu de « composants ». Nous présenterons les composants dans la [prochaine section](/docs/components-and-props.html). Les éléments représentant la base des composants, nous vous encourageons à bien lire cette section avant d’aller plus loin.

> ## Rendu d’un Élément dans le DOM {#rendering-an-element-into-the-dom}

Reviewer: Alors oui sauf qu'on fuit comme la peste « rendre » dans le contexte de ces docs. Circonlocutions donc, style « Faire le rendu » ou « Produire le rendu », ou simplement « afficher » (tout comme je favorise « rafraîchissement » pour “re-render(ing)”)

> Disons qu’il y ait une `<div>` quelque part dans votre fichier HTML :

Reviewer 2: "une balise `<div>`"

Les anglais on l’habitude de transformer en nom ou en verbe un peu tout et n'importe quoi. S'il est vraie que c'est une abus de langage fréquent chez les tech français du web de dire "une `<div>`", ça reste un abus de langage d'utiliser `<div>` comme un nom et dans une doc technique ça vaut le cout d'être un peu rigoureux sur la qualification des choses ;)

```html
<div id="root"></div>
```

> Nous l’appelons nœud DOM « racine » car tout ce qu’il contient sera géré par le DOM React.

Reviewer 2: "Nous l’appelons nœud DOM..." -> "Nous appelons ça un nœud DOM..."

> 

Reviewer 2: Tu as oublié de traduire cette phrase ;)

Les applications dévéloppées uniquement avec React ont généralement un seul nœud DOM racine. Si vous intégrez React dans une application existante, vous pouvez avoir autant de nœuds DOM racines isolés que vous le souhaitez.

`embed:rendering-elements/render-an-element.js`

[](codepen://rendering-elements/render-an-element)

> Cela affiche « Bonjour, monde » sur la page.

Reviewer 2: "Cela affiche" -> "Ça affiche"
cf. : https://github.com/reactjs/fr.reactjs.org/wiki/Meilleures-pratiques-de-traduction#ton

Si tu tiens à utiliser "cela", j'aurai alors tendance à privilégier un future plutôt qu'un présent (ça réduit un peu la charge formel de la phrase) :
"Ça affiche « Bonjour, monde » sur la page."
"Cela affichera « Bonjour, monde » sur la page."

Après dans ce contexte, "Ça" sonne un peu moche et j'aurai spontanément évité cet article indéfini en le recontextualisant : "Cet exemple de code affichera « Bonjour, monde » sur la page."

## Mettre à Jour un Élément Rendu {#updating-the-rendered-element}

Les éléments React sont [immuables](https://fr.wikipedia.org/wiki/Objet_immuable). Une fois votre élément créé, vous ne pouvez plus modifier ses enfants ou ses attributs. Un élément est comme une image d’un film à un instant T : il représente l’interface utilisateur à un point précis dans le temps.

Avec nos connaissances actuelles, la seule façon de mettre à jour l’interface utilisateur est de créer un nouvel élément et de le passer à `ReactDOM.render()`.

> Prenons l’exemple de l’horloge :

Reviewer 2: "Prenons l’exemple de l’horloge" -> "Prenons l'exemple d'une horloge" ou plus proche de l'original "Prenons l'exemple de cette horloge"

`embed:rendering-elements/update-rendered-element.js`

[](codepen://rendering-elements/update-rendered-element)

> À chaque seconde, nous appellons `ReactDOM.render()` depuis la fonction de rappel [`setInterval()`](https://developer.mozilla.org/en-US/docs/Web/API/WindowTimers/setInterval).

Reviewer 2: Il y a un léger contre sens là, `setInterval()` n'est pas une fonction de rappelle, c'est une API qui appelle une fonction à intervalle régulier :

"À chaque seconde, nous appelons `ReactDOM.render()` depuis une fonction de rappel passée à `setInterval()`"

>**Note :**
>
>En pratique, la plupart des applications React n’appellent `ReactDOM.render()` qu’une seule fois. Dans les sections à venir, nous apprendrons comment encapsuler un tel code dans des [composants à état](/docs/state-and-lifecycle.html).
>

> >Nous vous conseillons de ne pas sauter ces sujets car ils sont dépendants les uns des autres.

Reviewer 2: "Nous vous conseillons de lire les sujets dans l'ordre car ils sont dépendants les uns des autres."

> ## React Met à Jour Seulement ce Qui Est Nécessaire {#react-only-updates-whats-necessary}

Reviewer 2: Ça ne m'avait pas sauté aux yeux sur les autres titres, mais la capitalisation des mots des titres est un truc spécifiquement anglais. En français on ne capitalise que la première lettre du premier mot (comme une phrase)

Author: En effet 🙂 
Par contre dans les meilleures pratiques de traduction, il a une partie sur [la capitalisation des titres](https://github.com/reactjs/fr.reactjs.org/wiki/Meilleures-pratiques-de-traduction#titres-ids-de-titres-et-ancres-des-liens). Donc à voir si on les retire partout et on modifie le wiki en conséquence (car il me semble que sur les autres PR aussi, les titres sont capitalisés)

Reviewer 2: Ah, bien vu... faut que j'en parle aux autre mainteneur parce que je ne suis pas très chaud pour la capitalisation des titres. Laisse comme ça pour l'instant et on avisera si on change la recommandation.

Author: Ça roule, merci pour cette review ! J'ai push les corrections

Reviewer: L'anglais en soi ne capitalise pas les titres, ça vient de la presse, spécifiquement, pas de la littérature. Je pense que c'est un choix esthétique de la part de la team React et qu'ils veulent la préserver dans les autres langues romanes, mais on peut clairement demander à Dan (je fais ça illico), auquel cas je ferai plus tard une PR dédiée à a recapitalisation des contenus déjà traduits et fusionnés.

Le DOM React compare un élément et ses enfants avec la version précédente, et applique uniquement les mises à jour nécessaires au DOM pour que celui-ci reflète bien l’état voulu.

> Vous pouvez vérifier ce comportement en inspectant le [dernier example](codepen://rendering-elements/update-rendered-element) avec les outils de développement du navigateur :

Reviewer 2: example (en) -> exemple (fr) ;)

> ![L’inspecteur montrant les mises à jour périodiques](../images/docs/granular-dom-updates.gif)

Reviewer 2: granular != périodique
granular == atomique

L’inspecteur montrant des mises à jour atomiques

Même si nous créons chaque seconde un élément décrivant l’arborescence complète de l’interface utilisateur, seul le nœud texte dont le contenu a été modifié est mis à jour par le DOM React.

> Par expérience, réfléchir à quoi devrait ressembler une interface utilisateur à un moment donné plutôt qu’à comment elle devrait évoluer dans le temps permet d’éliminer toute une classe de bugs.

Reviewer 2: "L’expérience nous montre que réfléchir à quoi devrait ressembler une interface utilisateur à un moment donné plutôt que de réfléchir à comment elle devrait évoluer permet d'éliminer tout un ensemble de bugs."

Reviewer: Je plussoie Jérémie à ceci près qu’au lieu d’« un ensemble », j’utiliserais « une catégorie »

---
id: rendering-elements
title: Le rendu des éléments
permalink: docs/rendering-elements.html
redirect_from:
  - "docs/displaying-data.html"
prev: introducing-jsx.html
next: components-and-props.html
---

> Les éléments constituent les plus petits blocs d’une application React.

Reviewer: ```suggestion
Les éléments sont les blocs élémentaires d’une application React.
```

Allègement

> Un élément retranscrit ce que vous voulez voir à l’écran :

Reviewer: ```suggestion
Un élément décrit ce que vous voulez voir à l’écran :
```

Pourquoi changer le verbe, quand l'original marche bien ? Par ailleurs, on essaie de garder « retranscrire » pour le processus de réconciliation, qui passe des éléments React au DOM.

```js
const element = <h1>Bonjour, monde</h1>;
```

> Contrairement aux éléments DOM d’un navigateur, les éléments React sont de simples objets peu coûteux à créer. Le DOM React se charge de mettre à jour le DOM afin qu’il corresponde aux éléments React.

Reviewer: ```suggestion
Contrairement aux éléments DOM d’un navigateur, les éléments React sont de simples objets peu coûteux à créer. React DOM se charge de mettre à jour le DOM afin qu’il corresponde aux éléments React.
```

Piège ! 😨  L’expression « React DOM » fait toujours référence au module du même nom, qui est le réconcilieur de React pour la couche DOM du navigateur. Toujours laisser tel quel 😉

> >**Note :**

Reviewer: ```suggestion
>**Remarque :**
```

cf. [Glossaire](https://github.com/reactjs/fr.reactjs.org/wiki/Glossaire)

>

> >On pourrait confondre les éléments avec un concept plus largement connu de « composants ». Nous présenterons les composants dans la [prochaine section](/docs/components-and-props.html). Les éléments représentant la base des composants, nous vous encourageons à bien lire cette section avant d’aller plus loin.

Reviewer: ```suggestion
>On pourrait confondre les éléments avec le concept plus répandu de « composants ». Nous présenterons les composants dans la [prochaine section](/docs/components-and-props.html). Les éléments représentant la base des composants, aussi nous vous encourageons à bien lire cette section avant d’aller plus loin.
```

C'est bien top, ça, notamment la reformulation « représentent la base », GG.  Juste quelques retouches d'allègement.

## Afficher un élément dans le DOM {#rendering-an-element-into-the-dom}

> Disons qu’il y ait une balise `<div>` quelque part dans votre fichier HTML :

Reviewer: ```suggestion
Supposons qu’il y ait une balise `<div>` quelque part dans votre fichier HTML :
```

Le « disons » rend assez bizarre, surtout parce qu'il indique généralement un fait avéré, qui entraînerait l'indicatif.

```html
<div id="root"></div>
```

> Nous appelons ça un nœud DOM « racine » car tout ce qu’il contient sera géré par le DOM React.

Reviewer: ```suggestion
Nous appelons ça un nœud DOM « racine » car tout ce qu’il contient sera géré par React DOM.
```

Pu— je sens que j'ai pas fini 😁

Les applications dévéloppées uniquement avec React ont généralement un seul nœud DOM racine. Si vous intégrez React dans une application existante, vous pouvez avoir autant de nœuds DOM racines isolés que vous le souhaitez.

> Pour faire le rendu d’un élément React dans le nœud DOM racine, indiquez-les à la méthode `ReactDOM.render()` :

Reviewer: ```suggestion
Pour faire le rendu d’un élément React dans un nœud DOM racine, passez les deux à la méthode `ReactDOM.render()` :
```

Petits tweaks…

`embed:rendering-elements/render-an-element.js`

> [](codepen://rendering-elements/render-an-element)

Reviewer: ```suggestion
**[Essayer sur CodePen](codepen://rendering-elements/render-an-element)**
```

Voir [les MP](https://github.com/reactjs/fr.reactjs.org/wiki/Meilleures-pratiques-de-traduction#liens-codepen) pour les détails.

Cet exemple de code affichera « Bonjour, monde » sur la page.

> ## Mettre à jour un élément rendu {#updating-the-rendered-element}

Reviewer: ```suggestion
## Mettre à jour un élément affiché {#updating-the-rendered-element}
```

On évite vraiment « rendre » comme verbe (y compris ici, où c'est le participe passé et pas un nom)

Les éléments React sont [immuables](https://fr.wikipedia.org/wiki/Objet_immuable). Une fois votre élément créé, vous ne pouvez plus modifier ses enfants ou ses attributs. Un élément est comme une image d’un film à un instant T : il représente l’interface utilisateur à un point précis dans le temps.

Avec nos connaissances actuelles, la seule façon de mettre à jour l’interface utilisateur est de créer un nouvel élément et de le passer à `ReactDOM.render()`.

Prenons l’exemple de cette horloge :

`embed:rendering-elements/update-rendered-element.js`

> [](codepen://rendering-elements/update-rendered-element)

Reviewer: ```suggestion
**[Essayer dans CodePen](codepen://rendering-elements/update-rendered-element)**
```

À chaque seconde, nous appellons `ReactDOM.render()` depuis une fonction de rappel passée à [`setInterval()`](https://developer.mozilla.org/fr/docs/Web/API/WindowTimers/setInterval).

> >**Note :**

Reviewer: ```suggestion
>**Remarque :**
```

>

> >En pratique, la plupart des applications React n’appellent `ReactDOM.render()` qu’une seule fois. Dans les sections à venir, nous apprendrons comment encapsuler un tel code dans des [composants à état](/docs/state-and-lifecycle.html).

Reviewer: ```suggestion
>En pratique, la plupart des applications React n’appellent `ReactDOM.render()` qu’une seule fois. Dans les prochaines sections, nous apprendrons comment encapsuler un tel code dans des [composants à état](/docs/state-and-lifecycle.html).
```

>

> >Nous vous conseillons de lire les sujets dans l’ordre car ils sont dépendants les uns des autres.

Reviewer: ```suggestion
>Nous vous conseillons de lire les sujets abordés dans l’ordre car ils s'appuient l’un sur l’autre.
```

Je trouve que ça donne moins l'impression d'une réciprocité (et plus d'une progression) que la formulation d'origine, mais c'est un peu subjectif…

> ## React met à jour seulement ce qui est nécessaire {#react-only-updates-whats-necessary}

Reviewer: ```suggestion
## React met à jour le strict nécessaire {#react-only-updates-whats-necessary}
```

Je trouve que ça insiste encore plus sur la recherche d'optimalité (et c'est plus fluide)

> Le DOM React compare un élément et ses enfants avec la version précédente, et applique uniquement les mises à jour nécessaires au DOM pour que celui-ci reflète bien l’état voulu.

Reviewer: ```suggestion
React DOM compare l’élément et ses enfants avec la version précédente, et applique uniquement les mises à jour DOM nécessaires pour refléter l’état voulu.
```

Vous pouvez vérifier ce comportement en inspectant le [dernier exemple](codepen://rendering-elements/update-rendered-element) avec les outils de développement du navigateur :

> ![L’inspecteur montrant des mises à jour atomiques](../images/docs/granular-dom-updates.gif)

Reviewer: GG le « atomique » suggéré par @JeremiePat, c'est nettement plus efficace ici que « granulaire » 💯

> Même si nous créons chaque seconde un élément décrivant l’arborescence complète de l’interface utilisateur, seul le nœud texte dont le contenu a été modifié est mis à jour par le DOM React.

Reviewer: ```suggestion
Même si nous créons chaque seconde un élément décrivant l’arborescence complète de l’interface utilisateur, seul le nœud texte dont le contenu a été modifié est mis à jour par React DOM.
```
